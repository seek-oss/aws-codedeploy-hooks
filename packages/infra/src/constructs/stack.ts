import { Stack, aws_iam } from 'aws-cdk-lib';
import * as nodejs from 'aws-cdk-lib/aws-lambda-nodejs';
import type { Construct } from 'constructs';

import { LAMBDA_HOOK_PROPS } from './lambda';

export type HookStackProps = Record<string, never>;

export class HookStack extends Stack {
  constructor(scope: Construct, id?: string, _props: HookStackProps = {}) {
    super(scope, id ?? 'HookStack', {
      description: 'AWS CodeDeploy hooks',
      stackName: 'aws-codedeploy-hooks',
      // TODO: set tags
      tags: undefined,
      terminationProtection: true,
    });

    const beforeAllowTrafficHook = new nodejs.NodejsFunction(
      this,
      'BeforeAllowTrafficHook',
      {
        ...LAMBDA_HOOK_PROPS,
        description: 'BeforeAllowTraffic hook deployed outside of a VPC',
        functionName: 'aws-codedeploy-hook-BeforeAllowTraffic',
        vpc: undefined,
      },
    );

    beforeAllowTrafficHook.addToRolePolicy(
      new aws_iam.PolicyStatement({
        actions: [
          'codedeploy:GetApplicationRevision',
          'codedeploy:GetDeployment',
          'codedeploy:PutLifecycleEventHookExecutionStatus',
          'lambda:InvokeFunction',
        ],
        resources: ['*'],
      }),
    );
  }
}
