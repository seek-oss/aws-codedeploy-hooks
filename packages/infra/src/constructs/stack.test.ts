import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

import { HookStack } from './stack';

it('returns expected CloudFormation stack', () => {
  const app = new App();

  const stack = new HookStack(app);

  const template = Template.fromStack(stack);

  template.resourceCountIs('AWS::Lambda::Function', 1);

  template.resourcePropertiesCountIs(
    'AWS::Lambda::Function',
    { VpcConfig: {} },
    0,
  );

  const json = JSON.stringify(template.toJSON()).replaceAll(
    /"S3Key":"([0-9a-f]+)\.zip"/g,
    (_, hash) => `"S3Key":"${'x'.repeat(hash.length)}.zip"`,
  );

  expect(JSON.parse(json)).toMatchSnapshot();
});
