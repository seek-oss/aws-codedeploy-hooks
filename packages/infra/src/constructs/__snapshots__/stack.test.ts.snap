// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns expected CloudFormation stack 1`] = `
{
  "Description": "AWS CodeDeploy hooks",
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "BeforeAllowTrafficHookFCD1BC97": {
      "DependsOn": [
        "BeforeAllowTrafficHookServiceRoleDefaultPolicyDB153C72",
        "BeforeAllowTrafficHookServiceRole78A95A40",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.zip",
        },
        "Description": "BeforeAllowTraffic hook deployed outside of a VPC",
        "Environment": {
          "Variables": {
            "NODE_ENV": "production",
            "NODE_OPTIONS": "--enable-source-maps",
          },
        },
        "FunctionName": "aws-codedeploy-hook-BeforeAllowTraffic",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "BeforeAllowTrafficHookServiceRole78A95A40",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BeforeAllowTrafficHookServiceRole78A95A40": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "BeforeAllowTrafficHookServiceRoleDefaultPolicyDB153C72": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "codedeploy:GetApplicationRevision",
                "codedeploy:GetDeployment",
                "codedeploy:PutLifecycleEventHookExecutionStatus",
                "lambda:InvokeFunction",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "*",
              "Condition": {
                "Null": {
                  "aws:ResourceTag/aws-codedeploy-hooks": "true",
                },
              },
              "Effect": "Deny",
              "Resource": "*",
            },
            {
              "Action": "*",
              "Condition": {
                "StringEquals": {
                  "aws:ResourceTag/aws-codedeploy-hooks": [
                    "",
                    "false",
                  ],
                },
              },
              "Effect": "Deny",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BeforeAllowTrafficHookServiceRoleDefaultPolicyDB153C72",
        "Roles": [
          {
            "Ref": "BeforeAllowTrafficHookServiceRole78A95A40",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
